gcovr/configuration.py

```diff
@@ -616,6 +616,12 @@ def merge_options_and_set_defaults(partial_namespaces, all_options=None):
         type=OutputOrDefault,
         default=None,
     ),
+    GcovrConfigOption(
+        "log_summary", ["--log_summary-uncovered"],
+        group="output_options",
+        help="Instead using stdout only, the summary uses logging as well.",
+        action="store_false",
+    ),
```

Comment:
The summary is a output file and should not be logged.
```suggestion
```

Reply:
I'm not able to access stdout in my project. How should I reroute this instead? Provide an output file instead?

Reply:
Why are you not able to access stdout? You can also redirect it to a file in the shell.

Reply:
While in theory I can encapsulate the call of our python script with a redirection into a file, since we not just log lines to our log files, but to stdout in general, this redirection would capture the entire content of the dedicated log file in addition to the summary. Which means that piping the contents of the redirection file into the log file leads to needless duplication. Copying that redirection file as secondary output causes just confusion. Removing the duplication somehow makes more effort and is going against how we deal with logging in general. So in worst case we'd rather patch that part locally, if we upgrade later.

But I don't get your opposition to this approach. This is hidden behind a flag. And if it is active then you still get it in the console, since the logger is configured to output to stdout. In our case, we would be able to reconfigure the logger ourselves to use the same log file as the rest of the script. So what is a better solution that can satisfy us both?

Reply:
If the logger always goes to `stderr` a simply redirection should be enough.

Reply:
Gcovr is a Unix-ish command line tool, which means that output will be generated on STDOUT and messages/warnings/errors are shown on STDERR. I'm fine with all the warnings being handled via standard logging, but I'm not OK with normal output of our reports taking this more fragile route. And the --print-summary output is effectively a kind of report.

I want to enable users to work effectively with gcovr. But I have to balance that against complexity and future maintenance effort. “This is hidden behind a flag” is not a strong argument for adding a feature. To me as a maintainer this sounds like “more complexity for marginal benefit”.

Eric Gunnerson has written a [great article about managing complexity on the C# compiler](https://docs.microsoft.com/en-us/archive/blogs/ericgu/minus-100-points). Core quotes:

> I've been spending a lot of time answering customer questions about why we do (or don't) have certain language features in C#. […]
>
> One of the big reasons we didn't do this is that it's really hard to remove complexity when you take a subtractive approach, as removing a feature in one area may not allow you to revisit low-level design decisions, nor will it allow you to remove complexity elsewhere, in places where it support the now-removed feature.
>
> So, we decided on the additive approach instead, and worked hard to keep the complexity down. One way to do that is through the concept of “minus 100 points”. Every feature starts out in the hole by 100 points, which means that it has to have a significant net positive effect on the overall package for it to make it into the language. Some features are okay features for a language to have, they just aren't quite good enough to make it into the language.

Here, we have the “--print-summary output should optionally be handled via logging” feature. It starts with -100 points. You're suggesting that it would make using gcovr in your specific environment easier where you cannot use tools that use the normal Unix conventions. That's maybe +1 point from my perspective, but still leaves -99 points.

This would also be a feature that boxes us into a certain design decision that we can't change later. Specifically, gcovr has not made and will never make any guarantees about how the output or messages arrive at STDOUT or STDERR. Whether we use `logging` or not is purely an implementation detail. Adding this option would change that from an implementation detail to something that's part of the public interface of this software. I'm very conscious about that in order to avoid [xkcd 1172](https://xkcd.com/1172/) situations.

I'm still very happy with the idea to use standard logging because the old `Logger` class makes the code unnecessarily complicated by having to pass around a logger object explicitly.

Reply:
Ok, I'll remove that part then.